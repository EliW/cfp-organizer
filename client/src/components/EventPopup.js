import React from 'react';

export const EventPopup = ({ event }) => (
  <React.Fragment>
    <a href={event.url} target="_blank">{event.name}</a> - { event.date } <br/>
    { event.city }, { event.country } <br/>
    <a href={event.cfpUrl} target="_blank">CFP Closes on {event.cfpClose}</a>
  </React.Fragment>
);