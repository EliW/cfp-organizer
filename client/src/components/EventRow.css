.SvgButton {
  height: 30px;
  width: 30px;
  border: none;
  background: none;
  cursor: pointer;
  margin-right: 10px;
}

.SvgButton:focus {
  outline: 3px auto #0074d9;
}

.SvgButton::-moz-focus-inner {
  border: 0;
}

.SvgButton svg {
  outline: none;
}

.u-srOnly {
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.EventRow-button {
  border: none;
  display: block;
  font-size: 1em;
  width: 100%;
}
