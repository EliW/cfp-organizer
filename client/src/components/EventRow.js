import React from 'react';

export const EventRow = ({ className, rowClassName, event }) => (
  <div className={`${className} EventRow`}>
    <div className={rowClassName}><a href={event.url}>{event.name}</a></div>
    <div className={`${rowClassName} Table-cell--location`}>
      {event.location} 
      <span className='u-small'>{event.city}, {event.country}, {event.countryCode}, {event.continent}</span>
    </div>
    <div className={rowClassName}>{event.date}</div>
    <div className={rowClassName}><a href={event.cfpUrl}>{event.cfpClose}</a></div>
    <div className={`${rowClassName} u-small`}>{event.eventTags}</div>
  </div>
);