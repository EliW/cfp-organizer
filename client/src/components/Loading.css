.Loading {
  color: #666;
  margin: 6em;
  text-align: center;
}

.Loading-spinner {
  display: inline-block;
  position: relative;
  width: 64px;
  height: 64px;
}
.Loading-spinner div {
  position: absolute;
  width: 5px;
  height: 5px;
  background: #aaa;
  border-radius: 50%;
  animation: Loading 1.2s linear infinite;
}
.Loading-spinner .one {
  animation-delay: 0s;
  top: 29px;
  left: 53px;
}
.Loading-spinner .two {
  animation-delay: -0.1s;
  top: 18px;
  left: 50px;
}
.Loading-spinner .three {
  animation-delay: -0.2s;
  top: 9px;
  left: 41px;
}
.Loading-spinner .four {
  animation-delay: -0.3s;
  top: 6px;
  left: 29px;
}
.Loading-spinner .five {
  animation-delay: -0.4s;
  top: 9px;
  left: 18px;
}
.Loading-spinner .six {
  animation-delay: -0.5s;
  top: 18px;
  left: 9px;
}
.Loading-spinner .seven {
  animation-delay: -0.6s;
  top: 29px;
  left: 6px;
}
.Loading-spinner .eight {
  animation-delay: -0.7s;
  top: 41px;
  left: 9px;
}
.Loading-spinner .nine {
  animation-delay: -0.8s;
  top: 50px;
  left: 18px;
}
.Loading-spinner .ten {
  animation-delay: -0.9s;
  top: 53px;
  left: 29px;
}
.Loading-spinner .eleven {
  animation-delay: -1s;
  top: 50px;
  left: 41px;
}
.Loading-spinner .twelve {
  animation-delay: -1.1s;
  top: 41px;
  left: 50px;
}
@keyframes Loading {
  0%, 20%, 80%, 100% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.5);
  }
}
