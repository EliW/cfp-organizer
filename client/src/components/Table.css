.Table {
  padding-bottom: 2em;
  margin: 0 2em;
}

.Table-row {
  display: flex;
}

.Table-cell {
  flex: 1;
  border: 1px solid #ccc;
  padding: 1em 1.5em;
}

.Table-row--header {
  height: 100px;
}

.Table-cell--header {
  font-weight: 700;
}

.Table-row:nth-child(even) .Table-cell {
  background-color: #eee;
}

.Table-cell--location {
  display: flex;
  flex-direction: column;
}

@media (max-width: 1000px) {
  .Table-cell:last-child { 
    display: none; 
  }
}

.Sortable {
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  height: 30px;
}

.Sortable-unsorted {
  display: flex;
  flex-direction: column;
}

.Sortable-unsorted .Arrow {
  margin-bottom: 3px;
}

.Arrow {
  width: 0; 
  height: 0; 
  border-left: 10px solid transparent;
  border-right: 10px solid transparent;
}

.Arrow--up {
  border-bottom: 10px solid #999;
}

.Arrow--down {
  border-top: 10px solid #999;
}