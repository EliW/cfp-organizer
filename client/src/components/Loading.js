import React from 'react';
import './Loading.css';

export const Loading = () => (
  <React.Fragment>
    <div className="Loading">
      <div className="Loading-spinner">
        <div className="one"></div>
        <div className="two"></div>
        <div className="three"></div>
        <div className="four"></div>
        <div className="five"></div>
        <div className="six"></div>
        <div className="seven"></div>
        <div className="eight"></div>
        <div className="nine"></div>
        <div className="ten"></div>
        <div className="eleven"></div>
        <div className="twelve"></div>
      </div>
      <p>Loading...</p>
    </div>
  </React.Fragment>
);