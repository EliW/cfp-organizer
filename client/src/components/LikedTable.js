import React from "react";
import { Table } from "./Table.js";

export const LikedTable = props => {
  const likedEventsFromStorage = localStorage.getItem(
    "cfpOrganizer-likedEvents"
  );
  let likedEvents = {};
  if (likedEventsFromStorage) {
    likedEvents = {
      ...JSON.parse(localStorage.getItem("cfpOrganizer-likedEvents"))
    };
  }

  const data = props.data.filter(event => likedEvents[event.name]);

  return <Table {...props} data={data} />;
};
