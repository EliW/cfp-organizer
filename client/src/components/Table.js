import React, { Component } from 'react';
import './Table.css';
import { TableHeader } from './TableHeader.js';

export class Table extends Component {
  state = {
    data: [],
    filteredData: [],
    filter: [], // [{ fields: [''], value: ''}]
    sort: {
      field: '',
      direction: ''
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.data && nextProps.data.length > 0) {
      this.setState({data: nextProps.data, filteredData: nextProps.data })
    }
  }

  compareBy = (key, sortAsc) => {
    let comparator;
    if (sortAsc) comparator = 1;
    if (!sortAsc) comparator = -1;

    return function (a, b) {
      let aValue = a[key];
      let bValue = b[key];
      if (typeof (aValue) === "string") aValue = aValue.toLowerCase();
      if (typeof (bValue) === "string") bValue = bValue.toLowerCase();

      if (aValue < bValue) return comparator * -1;
      if (aValue > bValue) return comparator * 1;
      return 0;
    };
  }

  sortBy = (sortKey) => {
    let arrayCopy = [...this.state.data];
    const sortAsc = this.state.sort.field === sortKey ? this.state.sort.direction !== 'asc' : true;
    arrayCopy.sort(this.compareBy(sortKey, sortAsc));

    this.setState({ filteredData: arrayCopy, sort: { field: sortKey, direction: sortAsc ? 'asc' : 'desc' } });
  }

  filterBy = (event, keys) => {
    let filters = [...this.state.filter]
    // remove existing filters for the keys
    filters = filters.filter(oldFilter => !oldFilter.fields.every(f => keys.includes(f)))

    if (event.target.value) {
      filters.push({ fields: keys, value: event.target.value.toLowerCase() })
    }

    function reducer(accumulator, filter) {
      accumulator = accumulator.filter((dataItem) => {
        return filter.fields.some(field => dataItem[field].toLowerCase().includes(filter.value))
      })
      return accumulator
    }

    const filteredData = filters.reduce(reducer, [...this.state.data])

    this.setState({ filteredData, filter: filters })
  }

  render() {
    return (
      <div className="Table">
        <header className="Table-row Table-row--header">
          { this.props.headers.map((header, i) => {
            return (
              <TableHeader
                key={i} 
                header={header}
                onSort={this.sortBy}
                onFilter={this.filterBy} 
                currentSort={this.state.sort}
              />
              )
            })
          }
        </header>
        <div className="Table-body">
         { this.state.filteredData.map((dataItem, i) => this.props.rowComponent(dataItem, i)) }
        </div>
      </div>
    )
  }
}