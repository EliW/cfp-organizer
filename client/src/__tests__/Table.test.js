/* eslint-env jest */
import * as React from 'react'
import moment from 'moment'
import { shallow } from 'enzyme'
import { Table } from '../components/Table'

let subject
let data

function loadSubject(props = {}) {
  data = [
  {
    name: 'name',
    url: 'http://url.com',
    date: 'September 30, 2018',
    momentDate: moment.utc('September 30, 2018', 'MMMM DD, YYYY'),
    cfpClose: 'September 30, 2018',
    cfpUrl: 'http://url.com',
    location: 'Chicago, IL',
    eventTags: 'Javascript, Ruby, Tech',
    city: 'Chicago',
    country: 'United States',
    countryCode: 'US',
    continent: 'North America',
    lat: 41,
    lng: -55,
  },
  {
    name: 'a name',
    url: 'http://url.com',
    date: 'October 30, 2018',
    momentDate: moment.utc('October 30, 2018', 'MMMM DD, YYYY'),
    cfpClose: 'October 30, 2018',
    cfpUrl: 'http://url.com',
    location: 'Philadelphia, PA',
    eventTags: 'Python, React',
    city: 'Philadelphia',
    country: 'United States',
    countryCode: 'US',
    continent: 'North America',
    lat: 41,
    lng: -40,
  }]

  subject = shallow(<Table
    data={data}
    headers={[]}
    rowComponent={jest.fn()}
    {...props}
  />)
}

describe('render', () => {
  it('renders properly', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })
})

describe('compareBy', () => {
  loadSubject()
  expect(subject.instance().compareBy('name', true)).toBeInstanceOf(Function)
})

describe('sortBy', () => {
  describe('when name asc', () => {
    it('sorts \'a name\' before \'name\'', () => {
      loadSubject()
      subject.instance().sortBy('name')
      expect(subject.state('filteredData')[0]).toEqual(data[1])
      expect(subject.state('filteredData')[1]).toEqual(data[0])
      expect(subject.state('sort')).toMatchObject({
        field: 'name',
        direction: 'asc'
      })
    })
  })

  describe('when name desc', () => {
    it('sorts \'name\' before \'a name\'', () => {
      loadSubject()
      subject.instance().setState({ sort: { field: 'name', direction: 'asc'} })
      subject.update()
      subject.instance().sortBy('name')
      expect(subject.state('filteredData')[0]).toEqual(data[0])
      expect(subject.state('filteredData')[1]).toEqual(data[1])
      expect(subject.state('sort')).toMatchObject({
        field: 'name',
        direction: 'desc'
      })
    })
  })

  describe('when momentDate asc', () => {
    it('sorts Sept 30 before Oct 30', () => {
      loadSubject()
      subject.instance().sortBy('momentDate')
      expect(subject.state('filteredData')[0]).toEqual(data[0])
      expect(subject.state('filteredData')[1]).toEqual(data[1])
      expect(subject.state('sort')).toMatchObject({
        field: 'momentDate',
        direction: 'asc'
      })
    })
  })
})

describe('filterBy', () => {
  describe('filter by location', () => {
    it('returns events in Europe', () => {
      loadSubject()
      subject.instance().filterBy({ target: { value: 'Europe' } }, ['location', 'country', 'continent', 'city', 'countryCode'])
      expect(subject.state('filteredData').length).toEqual(0)
      expect(subject.state('filter').length).toEqual(1)
      expect(subject.state('filter')[0].fields).toEqual(expect.arrayContaining(['location', 'country', 'continent', 'city', 'countryCode']))
    })
  })

  describe('filter by location and tags', () => {
    it('returns events in United States and tagged with Javscript', () => {
      loadSubject()
      subject.instance().filterBy({ target: { value: 'united states' } }, ['location', 'country', 'continent', 'city', 'countryCode'])
      subject.instance().filterBy({ target: { value: 'javascript' } }, ['eventTags'])
      expect(subject.state('filteredData').length).toEqual(1)
      expect(subject.state('filteredData')[0]).toEqual(data[0])
      expect(subject.state('filter').length).toEqual(2)
      expect(subject.state('filter')[1].fields).toEqual(expect.arrayContaining(['eventTags']))

    })
  })
})