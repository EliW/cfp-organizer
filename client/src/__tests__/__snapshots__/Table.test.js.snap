// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Table
    data={
        Array [
            Object {
              "cfpClose": "September 30, 2018",
              "cfpUrl": "http://url.com",
              "city": "Chicago",
              "continent": "North America",
              "country": "United States",
              "countryCode": "US",
              "date": "September 30, 2018",
              "eventTags": "Javascript, Ruby, Tech",
              "lat": 41,
              "lng": -55,
              "location": "Chicago, IL",
              "momentDate": "2018-09-30T05:00:00.000Z",
              "name": "name",
              "url": "http://url.com",
            },
            Object {
              "cfpClose": "October 30, 2018",
              "cfpUrl": "http://url.com",
              "city": "Philadelphia",
              "continent": "North America",
              "country": "United States",
              "countryCode": "US",
              "date": "October 30, 2018",
              "eventTags": "Python, React",
              "lat": 41,
              "lng": -40,
              "location": "Philadelphia, PA",
              "momentDate": "2018-10-30T05:00:00.000Z",
              "name": "a name",
              "url": "http://url.com",
            },
          ]
    }
    headers={Array []}
    rowComponent={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="Table-row Table-row--header"
>
          
</header>,
        <div
          className="Table-body"
>
          
</div>,
      ],
      "className": "Table undefined",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "Table-row Table-row--header",
        },
        "ref": null,
        "rendered": Array [],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            undefined,
          ],
          "className": "Table-body",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          undefined,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="Table-row Table-row--header"
>
            
</header>,
          <div
            className="Table-body"
>
            
</div>,
        ],
        "className": "Table undefined",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "Table-row Table-row--header",
          },
          "ref": null,
          "rendered": Array [],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              undefined,
            ],
            "className": "Table-body",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            undefined,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
